### PR TITLE
Issue 132b - P2/P3 Adding Users to Secure Directories

### PR DESCRIPTION
**Pull request by jofeinstein**
- Created 4 new models in `coldfront/core/allocation/models.py`
    - `SecureDirAddUserRequestStatusChoice, SecureDirAddUserRequest, SecureDirRemoveUserRequestStatusChoice, SecureDirRemoveUserRequest`
- Created 4 new forms in `coldfront/core/allocation/forms_/secure_dir_forms.py`
- Created 5 new views in `coldfront/core/allocation/views_/secure_dir_views.py`
- Added management command `create_directory_defaults`
    - creates objects necessary to store information about secure directories in the database 
- Added 2 new utils in `coldfront/core/allocation/utils.py`
    - create_secure_dir: creates a secure directory attached to the given project with the given subdirectory_name
    - get_secure_dir_manage_user_request_objects: fetches secure directory user management object and stores them in the given object
- Added SecureDirAddUserRequest and SecureDirRemoveUserRequest to the "Request Hub"
- Allocation tab on the navbar is conditionally visible to managers, PIs, staff and admin
    - added allocation_navbar_visibility to `coldfront/core/utils/context_processors.py`

- The way SecureDirAddUserRequest and SecureDirRemoveUserRequest are shown and handled was very similar so the same views handled both
    - For example, `SecureDirManageUsersView` can show either users to add or users to remove
    - Used kwargs in url and get_secure_dir_manage_user_request_objects to fetch relevant objects based on kwargs
    - Similar to how we used a "completed" variable to use the same view to show pending/completed requests

- Added tests for new views and utils
    - views: `python manage.py test coldfront.core.allocation.tests.test_views.test_secure_dir_views`
        - 87% coverage of secure_dir_views.py achieved
        - Mostly missed email exception messages and search filtering conditionals
    - utils: `python manage.py test coldfront.core.allocation.tests.test_utils.test_secure_dir_utils`
        - Near total coverage of 2 new utils. Only error checking not tested.

- `create_directory_defaults` needs to be run to generate relevant secure dir objects

### CODE REVIEW

**Comment by matthew-li** [comment]
@jofeinstein Is there a way to load in the existing directories (e.g., from a file)? We'll need to do this when deploying to production.

**Review by matthew-li — CHANGES_REQUESTED** [review]
The core functionality works as expected. Thanks for refactoring where possible and for extensively testing this.

I left mostly minor comments, though PI-handling in `create_secure_dir` might have downstream effects.

I also included `create_directory_defaults` (renamed) in Ansible and merged in `develop`, so you'll need to run the steps in #400 if you haven't already.

File: coldfront/core/allocation/forms_/secure_dir_forms.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    first_name = forms.CharField(max_length=30, required=False, disabled=True)
+    last_name = forms.CharField(max_length=150, required=False, disabled=True)
+    email = forms.EmailField(max_length=100, required=False, disabled=True)
+    selected = forms.BooleanField(initial=False, required=False)
+
+
+class SecureDirManageUsersSearchForm(forms.Form):
+    allocation_name = forms.CharField(label='Allocation Name',
+                                      max_length=100, required=False)
+    resource_name = forms.CharField(label='Directory Name',
```
[the remark this answers is not in the document]
Comment (jofeinstein) [reply coldfront/core/allocation/forms_/secure_dir_forms.py:15]: The way the reset buttons have worked in other search forms is that they just clear what's in the form. You have to press "search" again to get the non-filtered results

File: coldfront/core/allocation/management/commands/add_directory_defaults.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+class Command(BaseCommand):
+    help = 'Manually creates objects for storing relevant cluster directories.'
+    logger = logging.getLogger(__name__)
+
+    def handle(self, *args, **options):
+        """
+        Creates default objects used for storing cluster directories
+        """
+        from coldfront.core.resource.models import AttributeType
```
[the remark this answers is not in the document]
Comment (jofeinstein) [reply coldfront/core/allocation/management/commands/add_directory_defaults.py:25]: I put the imports inside the code bc there are two types of "AttributeType". One from resource and one from allocation.

File: coldfront/core/allocation/utils.py [hunk 3]
```diff
@@ -238,3 +250,123 @@ def review_cluster_access_requests_url():
     domain = settings.CENTER_BASE_URL
     view = reverse('allocation-cluster-account-request-list')
     return urljoin(domain, view)
+
+
+def create_secure_dir(project, subdirectory_name):
+    """
+    Creates two secure directory allocations: group directory and
```
[the remark this answers is not in the document]
Comment (jofeinstein) [reply coldfront/core/allocation/utils.py:257]: I took out the creation of AllocationUsers from the function. We can create a new func when we deal with requests.

File: coldfront/core/allocation/views_/secure_dir_views.py [hunk 4]
```diff
@@ -0,0 +1,755 @@
...
+                                          UserPassesTestMixin,
+                                          ListView):
+    template_name = 'secure_dir/secure_dir_manage_user_request_list.html'
+    paginate_by = 30
+
+    def test_func(self):
+        """UserPassesTestMixin tests."""
+        if self.request.user.is_superuser:
+            return True
+
```
[the remark this answers is not in the document]
Comment (jofeinstein) [reply coldfront/core/allocation/views_/secure_dir_views.py:316]: Users with view_securediradduserrequest and view_securedirremoveuserrequest can see these. I'll update the staff permissions to include these

**Comment by jofeinstein** [comment]
I made the requested changes.

**Comment by jofeinstein** [comment]
I added permissions to the staff group so `python manage.py create_staff_group` will have to be rerun on deployment

File: coldfront/core/allocation/management/commands/add_directory_defaults.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (matthew-li) [reply coldfront/core/allocation/management/commands/add_directory_defaults.py:25]: Got it--one way to get around this is to `import X as Y`, but this is fine.

**Comment by matthew-li** [comment]
@jofeinstein Some classes in `secure_dir_views` are missing loggers, causing exceptions when trying to log that a notification email couldn't send.

Would recommend just having one logger at the top-level.

File: coldfront/core/user/templates/request_hub/request_hub.html [hunk 5]
```diff
@@ -63,6 +63,14 @@ <h1>Request Hub</h1>
     {% include 'request_hub/request_section.html' %}
```
[the remark this answers is not in the document]
Comment (matthew-li) [reply coldfront/core/user/templates/request_hub/request_hub.html:63]: Confirmed, thanks.

**Comment by matthew-li** [comment]
@jofeinstein Thanks for making the changes--things look good now. I ran into one other issue (above). Once resolved, I'll run the test suite and then merge.

Offline, can you provide me with a one-time script to import existing directories and their users? I'll just run it manually as below, since we're not going to need it again.
```
python manage.py shell < your_script_here.py
```

**Comment by jofeinstein** [comment]
Made the logger changes and I'll get that script to you soon

**Review by matthew-li — APPROVED** [review]
Tests are now passing. Looks good, thanks!

**Pull request merged**